src/lsm/set_associative_cache.zig

```diff
@@ -0,0 +1,423 @@
+                Value,
+                value_alignment,
+                value_count_max,
+                .exact,
+            );
+            errdefer allocator.free(values);
+
+            const counts = try allocator.alloc(u64, @divExact(counts_size, @sizeOf(u64)));
+            errdefer allocator.free(counts);
+
+            const clocks = try allocator.alloc(u64, @divExact(counts_size, @sizeOf(u64)));
+            errdefer allocator.free(clocks);
+
+            _ = Clocks;
+            _ = Counts;
+
+            var self = Self{
+                .sets = sets,
+                .tags = tags,
+                .values = values,
+                .counts = counts,
+                .clocks = clocks,
+            };
+
+            self.reset();
+
+            return self;
+        }
+
+        pub fn deinit(self: *Self, allocator: mem.Allocator) void {
+            assert(self.sets > 0);
+            self.sets = 0;
+
+            allocator.free(self.tags);
+            allocator.free(self.values);
+            allocator.free(self.counts);
+            allocator.free(self.clocks);
+        }
+
+        pub fn reset(self: *Self) void {
+            mem.set(Tag, self.tags, 0);
+            mem.set(u64, self.counts, 0);
+            mem.set(u64, self.clocks, 0);
+        }
+
+        pub fn get(self: *Self, key: Key) ?*align(value_alignment) Value {
+            const set = self.associate(key);
+            return self.search(set, key);
+        }
+
+        pub const GetOrPutResult = struct {
+            value_ptr: *align(value_alignment) Value,
+            found_existing: bool,
+        };
+
+        pub fn get_or_put(self: *Self, key: Key) GetOrPutResult {
+            return self.get_or_put_preserve_locked(
+                void,
+                struct {
+                    inline fn locked(_: void, _: *const Value) bool {
+                        return false;
+                    }
+                }.locked,
+                {},
+                key,
+            );
+        }
+
+        /// Get a pointer to an existing value with the given key if the key is in the cache.
+        /// If the key is not in the cache, add it to the cache, evicting older entries if needed.
+        /// Never evicts keys for which locked() returns true.
+        /// The caller must guarantee that locked() returns true for less than layout.ways keys.
+        pub fn get_or_put_preserve_locked(
+            self: *Self,
+            comptime Context: type,
+            comptime locked: fn (Context, *align(value_alignment) const Value) callconv(.Inline) bool,
+            context: Context,
+            key: Key,
+        ) GetOrPutResult {
+            const set = self.associate(key);
+
+            if (self.search(set, key)) |existing| {
+                assert(!locked(context, existing));
+                return .{
+                    .value_ptr = existing,
+                    .found_existing = true,
+                };
+            }
```

Comment:
When acquiring a block buffer from the cache, it feels like this is work that would have already been done upfront earlier at the call site?

i.e. `start_read()` only seems to need a `value_ptr` slot that hasn't been locked?

Reply:
Indeed, if we always remove the corresponding block from the cache when `free_set.release()` we don't need this search call here. We could keep it as an assertion if `verify = true` however.

I'm inclined to scrap the `get_or_put()` API for a `put_no_clobber()` API that assumes the inserted key does not already exist in the cache and asserts that fact if `lsm.verify = true`. I believe that would satisfy the requirements for the object cache as well. 

`get_or_put()` seemed like a nice idea as it's pretty much the most flexible API possible, but it seems we don't need that flexibility here :)